text/0015-Service-Reference-Retrieval.md

```diff
@@ -0,0 +1,232 @@
+- Start Date: 2024-10-10
+- CppMicroServices PR: https://github.com/CppMicroServices/CppMicroServices/pull/1044
+
+# Add an free API in that can retrieve a <code>ServiceObject</code>'s <code>ServiceReference</code> from itself
+
+## Summary
+
+Currently there is no API in the <code>cppmicroservices</code> namespace that can retrieve a <code>ServiceReference</code> from the <code>ServiceObject</code> itself.
+This document discusses need for such API in the particular use case of declarative services (DS) and its possible design/solution.
+
+## Motivation
+
+Developers using DS rely on DS's SCR (service component runtime) to inject references of services statically at construction or dynamically using the <code>bind</code> and <code>unbind</code> methods. These <code>ServiceObject</code>s can then be used by the client. However, clients often use properties in the Service to store metadata relevant to that service. Those properties, however, are not available to the <code>ServiceObject</code> itself, only via its <code>ServiceReference</code>.
+
+Given this it seems appropriate to have a new method that the client can use to translate a <code>ServiceObject</code> into its <code>ServiceReference</code>.
+
+## Requirements
+
+This API must:
+
+1. take in a <code>ServiceObject\<void\></code> and return the <code>ServiceReference\<void\></code> for that service
```

Comment:
isn't it supposed to take a `std::shared_ptr<T>`, where `T` is a service interface registered with the cppms framework?

Reply:
The second bullet point discusses that functionality. You can give it to void or not. But after what conor explained above, unless you specifically ask for the <void> return type, the compiler will auto fill in the templates. So yes, probably more worth just talking about bullet #2, or at least mention that first

Reply:
and also, void fits into T, which means we probably don't need bullet 1 at all...